### ISSUE TITLE
Implement a number-count animation for DORA-cards

### ISSUE THREAD

**Issue by e-for-eshaan** [middlewarehq/middleware#385]
## Why do we need this ?
we can add some animation to the cards' numbers on
![image](https://github.com/middlewarehq/middleware/assets/76566992/bd1d95e5-85da-45d0-ab58-e7a2ff49c2b8)

similar to this
![use-count-up](https://github.com/middlewarehq/middleware/assets/76566992/fbdc8976-3664-4db7-9b5b-93a7c60c1d16)

## Acceptance Criteria

- try not to introduce packages within the code base, writing optimized code will be appreciated
- It's best if you apply these animation for the `Deployment frequency` and `Change failure rate` first. These look farely straight forward
- using a component/custom-hook based approach is appreciated

- [ ] TODO 1
- [ ] TODO 2
- [ ] TODO 3

**Comment by jayantbh** [comment]
Hey @KartikChawla09, sure you can take this up.

**Comment by jayantbh** [comment]
I'm assigning this to you. In case that changes, let me know and I'll unassign you.

**Comment by KartikChawla09** [comment]
> I'm assigning this to you. In case that changes, let me know and I'll unassign you.

Thank you for assigning it to me. Can you have a look at #459 while I work on this ?

**Comment by jayantbh** [comment]
Hey @KartikChawla09, #459 was merged. 👏🏽 👏🏽 👏🏽 
Feel free to continue on this.

**Comment by KartikChawla09** [comment]
> Hey @KartikChawla09, #459 was merged. 👏🏽 👏🏽 👏🏽 Feel free to continue on this.

Yes! I am working on it. Will raise a PR ASAP.

**Comment by KartikChawla09** [comment]
Hey @e-for-eshaan @jayantbh .
![image](https://github.com/middlewarehq/middleware/assets/97456320/78182a92-1c3a-413f-9622-e6e64f467211)
The page is getting stuck on this loading animation for more than 5 minutes. It used to work previously but after fetching the latest commit it stopped to work.
I am attaching the logs below.
![image](https://github.com/middlewarehq/middleware/assets/97456320/6b88c4cc-07fa-42d9-a674-fcb0a81b80e5)
![image](https://github.com/middlewarehq/middleware/assets/97456320/5c822c22-e1a2-45d1-af24-dfb9c1d9da07)
Is this an issue with my local deployment ?

**Comment by KartikChawla09** [comment]
> Hey @KartikChawla09, this shouldn't have happened. Did you check if there is an older commit that seems to not have this issue?

The last commit I remember working correctly was [this one](https://github.com/middlewarehq/middleware/commit/f3d3850b0a8f04659cab6329a692558119d13b52). But maybe the issue is at my end only.  It would really help if someone could confirm that by checking on their local deployment.

**Comment by e-for-eshaan** [comment]
Hey @KartikChawla09 , are you still working on this?

**Comment by KartikChawla09** [comment]
> Hey @KartikChawla09 , are you still working on this?

Yes! I have implemented the counter as well but I am having some issues with my local deployment due to which I am not able to test the code locally. I will push the code when I am able to fix the issue. I have sent the error on  Slack as well.

**Comment by e-for-eshaan** [comment]
@KartikChawla09 been a while, how's this coming up, should I assign this to someone else?

**Comment by KartikChawla09** [comment]
There was some issue with my local deployment, as I mentioned I raised as issue on Slack as well but it didn't get fixed so I couldn't test my code.

import { useEffect, useState } from "react";

const easeInOutQuad = (t: number) => {
  return t < 0.5 ? 2 * t * t : -1 + (4 - 2 * t) * t;
};

const CounterAnimation = ({
  targetValue,
  duration,
}: {
  targetValue: number;
  duration: number;
}) => {
  const [count, setCount] = useState(0);

  useEffect(() => {
    const flag = Math.floor(targetValue) === targetValue;
    let start = 0;
    const iterations = 10;
    const time = duration / iterations;
    const startTime = performance.now();
    const counter = setInterval(() => {
      const elapsed = performance.now() - startTime;
      const progress = elapsed / duration;
      const easedProgress = easeInOutQuad(progress);
      start += 1;
      flag
        ? setCount(Math.floor(easedProgress * targetValue))
        : setCount(parseFloat((easedProgress * targetValue).toFixed(2)));
      if (start >= iterations) {
        clearInterval(counter);
        setCount(targetValue);
      }
    }, time);
  }, [duration, targetValue]);

  return <>{count}</>;
};

export default CounterAnimation;


This code snippet might help the person who takes up this issue!

**Comment by jayantbh** [comment]
Thanks @KartikChawla09 for your contribution, even though we couldn't sort out issues with running Middleware on an 8GB RAM system. Hopefully we'll get an idea of how to deal with that in the near future.

Unless you're able to see if Gitpod works out for you (which based on the last time we chatted about this, you'd hit some limits).

I've unassigned the issue for now, such that someone else can take it up.

**Comment by xyfer17** [comment]
@jayantbh please assign this issue to me, I'll would like to pick this up.

**Comment by jayantbh** [comment]
Done, @xyfer17

**Comment by xyfer17** [comment]
@jayantbh can you help me out, I'm unable to load the dora metrics getting error from api side:

Endpoint: `dora_metrics/fetchTeamDoraMetrics`

```
{
    "message": "<!doctype html>\n<html lang=en>\n<title>400 Bad Request</title>\n<h1>Bad Request</h1>\n<p>Invalid data: extra keys not allowed (path pr_filter[base_branches][0])</p>\n"
}
```

![Screenshot 2024-10-21 110314](https://github.com/user-attachments/assets/203b456d-e7b3-4538-835f-86d4ef6721e1)
![Screenshot 2024-10-21 114142](https://github.com/user-attachments/assets/96b6232d-9776-4b60-8d8c-b9a444eac9df)

**Comment by jayantbh** [comment]
I think this broke as part of a recent axios upgrade. Could you check if this works with an older commit?

**Comment by xyfer17** [comment]
@jayantbh have checked using git bisect it broke in the [commit](https://github.com/middlewarehq/middleware/commit/eee202710976773016613e874b2cff677431a757)  by axios version change
```
Updates `axios` from 0.26.1 to 0.28.0
```

**Comment by jayantbh** [comment]
Ah damn dependabot update. Let's see if we can revert that.

**Comment by xyfer17** [comment]
@jayantbh I have few questions on for the above ehancement:

1. Should the animation run only once (only on every page render), or do we need to reset or loop the animation under certain conditions?

2. on Deployment frequency the value always showing in whole number, can i show the animation over there in decimal  places or whole number.

**Comment by jayantbh** [comment]
1. If you want to keep it simple, then once per load, meaning when team changes or date changes or user navigates away and back to the page.
2. A better implementation might be to load it once at least 5 minutes after any of the above conditions have happened. This makes the animation happen after some time of inactivity instead of being more frequent and feeling noisy.

**Comment by xyfer17** [comment]
Can you Please take a look on the https://github.com/middlewarehq/middleware/pull/610 for the above issue.
Thanks :)
